### ISSUE TITLE
note regarding omitting skU from envelope and deriving it from rwd instead

### ISSUE THREAD

**Issue by stef** [cfrg/draft-irtf-cfrg-opaque#84]
in the source there's a commented out part `## Envelope considerations` which mentions that not even `skU` needs to be included in the credentials wrapped in the envelope, instead skU can be derived from `rwd`. there's a note that this part should be "brought back after updating". if this is indeed the case, do not forget to change 
> Applications MUST include skU in secret_credentials
in the {#protocol-messages} section.

**Comment by chris-wood** [comment]
Yeah, we need to remove that commented out text. I'll file a PR for that.

**Comment by stef** [comment]
why do you want to remove it instead of bringing it back? i think this part makes sense.

**Comment by chris-wood** [comment]
> i think this part makes sense.

I need to think about this more, though my sense is that the bandwidth saved is not be worth the additional complexity. The original I-D had this to say about the issue:

```
   Note (storage/communication efficient EnvU): It is possible to
   shorten EnvU by including PubS in ClrEnv and omitting SecEnv and any
   other value from EnvU.  In this case, the user's key PrivU is derived
   from RwdU; e.g., the value PAD in KEYS is used as a seed for a key
   generation procedure that outputs PrivU (the length of PAD will
   depend on the specific key generation procedure).  Such EnvU consists
   of Nonce, PubS and the HMAC value, resulting in less storage at the
   server and less communication from server to client.  The server can
   further minimize storage space by deriving per-user OPRF keys kU from
   a single global secret key using a PRF, and it can use the same pair
   (PrivS,PubS) for all its users.  In this case, the per-user OPAQUE
   storage consists of Nonce, PubU and HMAC(Khmac; PubS).  While it may
   be "tempting" to omit Nonce for space savings, this can lead to
   vulnerabilities (see below).
```

@hugokraw, what do you think?

**Comment by hugokraw** [comment]
I am sure that there are cases where such a savings will be meaningful and I am sure it will not be the typical case; I can't say much more than that. Frankly, the main reason I wanted to "document" this option is that in Bjorn Hasse's AuCPace paper, the amount of storage and/or communication was the only aspect on which AuCPace had some performance advantage. That paper compared against a very wasteful interpretation of what OPAQUE needed and I wanted to document the fact that one can actually be very minimal in envelope's size. I would prefer to make progress with the specification without entering this level of optimization but if there are potential users of OPAQUE for which these savings are important we could at least leave it as a remark.

**Comment by kevinlewi** [comment]
Is there a systematic / standardized way to generate a private key from a sequence of random bytes (RwdU), for all of the keypair types that we want to support?

If so, then I think it would be great to incorporate this optimization into our specification, especially since this will affect how test vectors are generated.

If not, then I think we are better off not addressing this optimization, unless there are compelling reasons for practictioners to minimize the extra 32 bytes that are being used for encoding skU.

**Comment by chris-wood** [comment]
> Is there a systematic / standardized way to generate a private key from a sequence of random bytes (RwdU), for all of the keypair types that we want to support?

It depends on the key pair. HPKE [defines specifies a couple derivation functions](https://cfrg.github.io/draft-irtf-cfrg-hpke/draft-irtf-cfrg-hpke.html#name-derivekeypair) for popular groups. But this isn't exhaustive. Right now, I'm leaning towards not including this optimization, but perhaps this is also something we should bring up to the list.

**Comment by hugokraw** [comment]
> Is there a systematic / standardized way to generate a private key from a sequence of random bytes (RwdU), for all of the keypair types that we want to support?

The universal way is to assume that the key generation procedure works on a string of random bits as input and therefore you can always produce that number of bits out of the RO applied to rwdU. This is simple for the case of EC keys but would not be so simple for an RSA private key (which requires sampling prime numbers).

**Comment by bytemare** [comment]
Hello,

can't rwdu be used as input to h2c's HashToScalar and serve as a private key? (This is how I understood OPAQUE the first time I read about it, but I may be missing something. But if two users share the same password and the server uses the same evaluation key in OPRF, both users end up with the same private key.)

**Comment by bytemare** [comment]
With the latest advancements here, I'd like to bring this back up.

Consideration to support RSA in the Sigma setup (or anything other that hasn't defined a secure deterministic key pair generation) made this a blocker.

Since the 'main' AKE now is 3DH and that the group used for 3DH[ is set to be the same as the one used in the OPRF](https://github.com/cfrg/draft-irtf-cfrg-opaque/issues/104#issuecomment-772114793), we can leverage hashing _rwdu_ to the field for the secret key (e.g. hash-to-scalar() in hash-to-curve terminology). This would nicely lower the complexity and size of the envelope construction.

I'm thinking about something like this:
Registration:
```
Client:
1. N = Unblind(blind, response.data)
2. y = Finalize(pwdU, N, "OPAQUEXX")
3. nonce = random(32)
4. rwdU = HKDF-Extract(nonce, Harden(y, params))
5. sku = HashToScalar( rwdu, "Opaque-HashToScalarSku")
6. pku = sku*Base
7. authKey = HKDF-Expand( rwdu, "AuthKey")
8. authTag = Hmac ( authkey, nonce || pku || pks )
9. envU = envelope(nonce, authTag)

C -> S : pku, envU
```

The envelope does not contain encrypted secret material anymore.

Then during authentication:
```
C -> S : oprf request, AKE request
S -> C : oprf response, AKE response, pks, envU
```

The client then does the same steps as during registration and checks the tag. Upon successful verification, uses sku, pku, and pks to proceed with the AKE.

**Comment by bytemare** [comment]
For cases where the AKE does not support deterministic key pair generation based on a seed, we could still use the envelope.

**Comment by stef** [comment]
i understand that this would be a 3rd kind of envelope config, besides base and custom-identifier. but i'm still very much in favor of this.

**Comment by chris-wood** [comment]
@stef, @bytemare: would you be able to propose this new envelope as a PR?

**Comment by stef** [comment]
sure

**Comment by stef** [comment]
hmmm interesting, in this case there is no need for an envelope. which brings up the question if there is no envelope, should there be a mode-flag sent as AAD or will all of this be implicit and the HMAC verified over a virtual packet consisting of the modeflag, the nonce and the server pubkey? i assume we still want the nonce in there for the authentication of the mode and the server pubkey. or do we drop the whole envelope thing altogether? in the latter i guess the km[23] auth hmac would serve as a deterrent to partitioning oracles, right?

**Comment by bytemare** [comment]
The draft has the following comment

**Comment by bytemare** [comment]
The draft has the following comment

> It is possible to dispense with encryption in the construction of envelope to
> obtain a shorter envelope (resulting in less storage at the server and less
> communication from server to client). The idea is to derive client_private_key from prk.
> However, for cases where client_private_key is not a random string of a given length, we
> define a more general procedure. Namely, what is derived from prk is a random
> seed used as an input to a key generation procedure that generates the pair
> (client_private_key, client_public_key). In this case, secret_credentials is empty and cleartext_credentials
> contains server_public_key. The random key generation seed is defined as
> HKDF-Expand(KdKey; info="KG seed", L)
> where L is the required seed length. We note that in this encryption-less
> scheme, the authentication still needs to be random-key robust which HMAC
> satisfies.

This is where my suggestion came from, reducing the envelope to only keep the nonce and authentication tag, so the client can verify before continuing the AKE.

**Comment by stef** [comment]
yes, i remember. but this fragment is from a time, when the envelope was made up of a secret and a cleartext env part. in the current spec the cleartext is not part of the envelope anymore, only of the authentication tag calculation. how the client takes possession of the server pubkey is not specified anymore. it might be sent along by the server, or it might be pre-shared. but i guess the point is - if we stick with this, that we should still do an hmac over the server pubkey even in this case. and if we do, we might want to do this including a nonce. and if we send a nonce we can also send a mode flag. but the question is. if this is necessary at at all, or can all of this be skipped and the security of the protocol not be weakened by doing so.

**Comment by bytemare** [comment]
I'm sorry, it seems the Github app has some issues, my comments keep disappearing. Here is my original comment:

# 

The draft has the following comment

> It is possible to dispense with encryption in the construction of envelope to
> obtain a shorter envelope (resulting in less storage at the server and less
> communication from server to client). The idea is to derive client_private_key from prk.
> However, for cases where client_private_key is not a random string of a given length, we
> define a more general procedure. Namely, what is derived from prk is a random
> seed used as an input to a key generation procedure that generates the pair
> (client_private_key, client_public_key). In this case, secret_credentials is empty and cleartext_credentials
> contains server_public_key. The random key generation seed is defined as
> HKDF-Expand(KdKey; info="KG seed", L)
> where L is the required seed length. We note that in this encryption-less
> scheme, the authentication still needs to be random-key robust which HMAC
> satisfies.

This is where my suggestion came from, reducing the envelope to only keep the nonce and authentication tag, so the client can verify before continuing the AKE.

**Comment by bytemare** [comment]
If it's possible to have no envelope at all and still keep the same level of security, that would be ideal (and I would love that!).The client would only send a public key to the server, but this is not authenticated, and an attacker could send their own public key in lieu of the real client's.  Also, the

**Comment by stef** [comment]
true, if the server pubkey is being sent along this could be a problem, but if the server pubkey is preshared somehow and authenticated out-of-band, like e.g. during registration then i guess not, but that also means (which in some cases is totally ok) that the client stores some "state" the server pubkey.
btw the clients long-term pubkey is being stored by the server upon registration, in a secured channel. so that cannot be replaced by an attacker.

**Comment by bytemare** [comment]
Ok, I'm switching to a computer, this bug is really annoying, I'm terribly sorry.
_(original message)_
#
If it's possible to have no envelope at all and still keep the same level of security, that would be ideal (and I would love that!).

The 'client's trace' would be very loosely coupled, with the client not knowing if it got the right OPRF evaluation. On registration, the public key the client sends back to the server would not be authenticated.

**Comment by stef** [comment]
> On registration, the public key the client sends back to the server would not be authenticated.

iiuc registration requires a secured channel, like tls.

**Comment by stef** [comment]
the client can still verify using km[23] an authentication tag.

**Comment by bytemare** [comment]
Since we have no encryption, is the nonce even necessary? We could do the following, and I'm not sure we'd loose a security property:
```
prk = HKDF-Extract(hardened, nil)
sk, pk = deriveKeyPair(prk)
authKey = HKDF-Expand(prk, "AuthKey", Nh)
authTag = hmac( authKey, pkc || pks )
envelope = authTag
```

If we went _envelope-less_:

In the AKE, the client computes `ikm` using the `skc` and `pks` long-term values. So one can assume that this is sufficient to proof possession of the right AKE mac keys, and thus verify KE2-mac and produce KE3-mac.

**Comment by bytemare** [comment]
@chris-wood what do you think?

If we went _envelope-less_, would the AKE macs be sufficient proofs that the client derived the right key pair?

**Comment by hugokraw** [comment]
This minimization of the protocol reminds me of this[ joke](https://eartstohear.wordpress.com/2015/09/01/fresh-fish-sold-here-daily/) .
Interestingly, your suggestion of getting rid of the whole envelope and building on the AKE authentication *may* work with 3DH (one would need to prove it). In the case of SIGMA, if enc_server_info is empty, the protocol breaks completely. It  allows for an offline dictionary attack after a single session with the attacker acting as man in the middle between server and client. 

However, I do not think that assuming the client has the server's public key would be the common case (indeed, one of the main advantages of PAKEs is that you do not assume PKI or any form of trusted server's public key). In this case, the server sends its public key  to the client which must be authenticated by the client. This will require an HMAC computed on pkU. So deriving skU directly from prk would not obviate the envelope, it will be the same as now except for contents.encrypted_creds being empty (hence also pseudorandom_pad being omitted).

**Comment by stef** [comment]
true, my whole context is 3dh.
however the last time i checked (since there is a lot changing still all the time i might be very wrong) the specification removed the part formerly known as "cleartext envelope", and left it unspecified how the pkS and optionally the id[US] are being recieved/stored/etc by the client. the only part afais that is specified is how to calculate the mac over the envelope and use the pkS/id[SU] as aad. so it leaves the door open for storing this info in the client. for example i can easily imagine a setup where the client is a javascript implementation provided by the server via https containing already the pkS hardcoded in the js. the current specification leaves that open to the implementation. if this is not something that is desired we should say so in the spec i guess.

**Comment by hugokraw** [comment]
I think that you can find scenarios where you'll have pkS cached or hardwired into the client (e.g., an app in your phone that you download from the apple store). I would not "prohibit" such configuration but I also don't want to build on it as a common case.

**Comment by bytemare** [comment]
I confirm, this suggestion is thought in the 3DH setup, as the document is now focused on 3DH.

I can agree that this situation may look like the joke you're thinking about, but I hope you're not taking any offense that we're making these suggestions that _make sense_. You're the expert here, you know best :)

Regarding the server key caching, building OPAQUE on it would indeed go against the very usefulness of aPAKEs, like you already mentioned in other comments. But it would be very useful to store the public key as a configuration parameter like the others (group and hash identifiers, etc.)

In the case we keep the envelope, would the nonce still be necessary, and the authentication tag suffice?

**Comment by hugokraw** [comment]
``I can agree that this situation may look like the joke you're thinking about, but I hope you're not taking any offense 
Not at all!

**Comment by hugokraw** [comment]
`I can agree that this situation may look like the joke you're thinking about, but I hope you're not taking any offense `

Not at all! No offense. To the contrary, these are good and non trivial  questions. They are welcome.

**Comment by hugokraw** [comment]
`In the case we keep the envelope, would the nonce still be necessary, and the authentication tag suffice?`

If you cache pkS (and assume it is authentic) and derive skU from prk, then it would seem that just applying HMAC to a fixed string (under auth_key) would suffice. Not sure if there is a use for the nonce in this case. If you really want to specify something like this in this document, I would need to think more about it.

**Comment by bytemare** [comment]
I was more thinking about the regular case, where we don't cache pkS. 

Like in this previous comment

> Since we have no encryption, is the nonce even necessary? We could do the following, and I'm not sure we'd loose a security property:
> ```prk = HKDF-Extract(hardened, nil)
> skc, pkc = deriveKeyPair(prk)
> authKey = HKDF-Expand(prk, "AuthKey", Nh)
> authTag = hmac( authKey, pkc || pks )
> envelope = authTag

**Comment by hugokraw** [comment]
> skc, pkc = deriveKeyPair(prk)
> authKey = HKDF-Expand(prk, "AuthKey", Nh)
> authTag = hmac( authKey, pkc || pks )
> envelope = authTag

It should be: 
skc, [pkc] = deriveKeyPair(prk)
authKey = HKDF-Expand(prk, "AuthKey", Nh)
authTag = hmac( authKey, pks )
envelope = (pks, authTag)

Here pkc is recomputed  from prk only if needed by the client (none of the protocols we consider need that and for the case of elliptic curve keys, computing pkc costs an exponentiation, so no need to compute if not essential) and it does not need to be authenticated under hmac.  
pks does need to be transported in the envelope.
It could be omitted if it is assumed that the client has it. Even in this case it would be ood to send authTag so the client can check if the stored pks is authentic.

**Comment by bytemare** [comment]
Doesn't the server need the client's pkc?

**Comment by bytemare** [comment]
I see. Only at registration, actually. During login, the client doesn't need its public key during login.

**Comment by chris-wood** [comment]
I'm supportive of this change under the condition that we have a *standardized* way to deterministically derive keys from a seed. HPKE provides a number of algorithms for NIST curves, which would suffice for the 3DH groups here. But it's not clear to me we want to rely on this. 

@kevinlewi, what's your take?

**Comment by bytemare** [comment]
Since we already rely on hash-to-curve for the OPRF, why could rely on its hash-to-field for the scalar

**Comment by chris-wood** [comment]
> Since we already rely on hash-to-curve for the OPRF, why could rely on its hash-to-field for the scalar

This is fine for prime order groups. I want to emphasize again that there may be other AKE instantiations that *do not* have such groups.

**Comment by bytemare** [comment]
In #104, I understand that the group in 3DH must match what we use for OPRF, which is defined using prime-order groups. I'm confused.

**Comment by chris-wood** [comment]
Yes, that's fine if we're using an AKE that is based on groups, like 3DH. That does not work if we're using an AKE based on, say, signatures, like SIGMA-I.

**Comment by bytemare** [comment]
I see. I was still fixed on using 3DH.

> The envelope is empty (and thus consumes no bytes on the wire).

See above

> Interestingly, your suggestion of getting rid of the whole envelope and building on the AKE authentication may work with 3DH (one would need to prove it). In the case of SIGMA, if enc_server_info is empty, the protocol breaks completely. It allows for an offline dictionary attack after a single session with the attacker acting as man in the middle between server and client.
> 
> However, I do not think that assuming the client has the server's public key would be the common case (indeed, one of the main advantages of PAKEs is that you do not assume PKI or any form of trusted server's public key). In this case, the server sends its public key to the client which must be authenticated by the client. This will require an HMAC computed on pkU. So deriving skU directly from prk would not obviate the envelope, it will be the same as now except for contents.encrypted_creds being empty (hence also pseudorandom_pad being omitted).

My takeaway is:
- > getting rid of the whole envelope and building on the AKE authentication may work with 3DH (one would need to prove it)

and we don't have that prove, yet.

- > the server sends its public key to the client which must be authenticated by the client. This will require an HMAC computed on pkU

So we still need an envelope, that would just be the authentication tag.

**Comment by kevinlewi** [comment]
After some thought and reviewing the comments here, I am not sure I am a big fan of adding the support for an extra envelope mode for this feature.

My understanding is that the main motivation for this third envelope mode is to reduce the size of the envelope, right? If it is a strictly better solution than the current proposal with "base mode", then I think we should aim to replace the existing "base mode" with one which incorporates these envelope optimizations, rather than introduce a third mode. Also, would there be an analogous version of this which supports custom identifiers but also with a smaller envelope size? Presumably we would not introduce a fourth new envelope mode, but instead this would simply be an improvement to the already-existing custom identifier mode as well.

Now, if we decide that the current proposal for base mode has properties that cannot be replicated with this compressed envelope proposal, then I think we should weigh the pros and cons of having those properties versus saving on the envelope bandwidth directly. How important is it that we minimize the size of the envelope / omit sending pkS across the wire, if it comes at the cost of dealing with extra assumptions that could end up being a footgun for the wrapping application?

Anyway, very interesting discussion. Looking forward to you all's thoughts on this!

**Comment by bytemare** [comment]
> My understanding is that the main motivation for this third envelope mode is to reduce the size of the envelope, right? If it is a strictly better solution than the current proposal with "base mode", then I think we should aim to replace the existing "base mode" with one which incorporates these envelope optimizations, rather than introduce a third mode.

Totally aligned with you, here. I was also thinking that it would be an interesting candidate to replace the base mode. The philosophy would then be
"Base mode envelope is for KeyGenFromSeed() compatible setups and doesn't store any secrets. Custom mode envelope offers encrypted secrets storage, that can be used for setups without a KeyGenFromSetup() function."

Other than reducing the size of the envelope, it slightly reduces the complexity of building the envelope and saves some calls to hashing.

> Also, would there be an analogous version of this which supports custom identifiers but also with a smaller envelope size?

Do you have an example in mind?

> How important is it that we minimize the size of the envelope / omit sending pkS across the wire, if it comes at the cost of dealing with extra assumptions that could end up being a footgun for the wrapping application?

Agreed that we need to be sure what we're doing here :)

**Comment by kevinlewi** [comment]
By the way, thinking more about the security of the Seed Mode proposal:

Base Mode:
- Client chooses skU from uniformly generated randomness

Seed Mode:
- Client derives skU directly from rwdU, which is in turn derived from (low) entropy of password and entropy of server's oprf_key.

One concern that I have with Seed Mode is that the client really has to trust that the server generated an independent oprf_key from fresh randomness in order to have confidence in its own skU. This is not an assumption that the client has to make in Base Mode.

To illustrate a hypothetical attack that would work on Seed Mode, but not on Base Mode: Imagine the server accidentally re-uses the same oprf_key for two clients, and these two clients happen to register with the same password. Then, their skUs will be identical, and in the AKE section, a *passive eavesdropper* may be able to learn things about the exchange that it wouldn't be able to learn if we had used Base Mode.

Of course, the above attack assumes that the server actually re-used the oprf_key... but the nice thing about Base Mode is that the re-use of the oprf_key there does not lead to this kind of passive eavesdropper attack... right?

**Comment by bytemare** [comment]
In seed mode, we still keep an authentication tag over the server's public key using a mac key derived from the same prk as for the secret key. So if the mac is valid, it means the chain of keys is valid (mac key < prk < hardened < unblinded).

Does that work?

**Comment by chris-wood** [comment]
Agreed, @kevinlewi. (This more strongly motivates the potential for this to be a separate mode, rather than a replacement for the base mode since analyzing it may take more time, which may ultimately slow things down.)

**Comment by bytemare** [comment]
As discussed with @kevinlewi this justifies the need to keep the nonce in the envelope.

**Comment by hugokraw** [comment]
It is true that in the seed mode you depend more on the quality of the OPRF key created on the server side. But in the base mode you depend entirely on the client randomness (if the client creates the key, which is what I expect to be the common case) and this will in many cases be of less quality than the server's.  For seed mode, the client should choose a random nonce that would be part of the seed derivation. While it is a public value, it still helps to randomize on top of the server's OPRF key. Also, this nonce will help against the case of a server that reuses the OPRF key with more than one user. Even if two users have the same password, the different nonces will produce different seeds.

Are we now involving the account name, say, user@server.com, in the derivation of (what we used to call) rwd?
Should we? Is there a well defined (deterministic and repeatable) identity string like that?

On the point of servers reusing the OPRF key with different users, there needs to be a HUGE warning in the RFC not to do that.   Btw, one of the mitigation of enumeration attacks uses a prf with userid name as input to derive the OPRF keys, so it ensures different oprf keys for different users.

**Comment by chris-wood** [comment]
Fixed in #84.
